hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/CalcDropLateItemsTransposeRule.java

```diff
@@ -0,0 +1,100 @@
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.opt.logical;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Calc;
+import org.apache.calcite.rel.rules.TransformationRule;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.immutables.value.Value;
+
+import java.util.List;
+
+import static com.hazelcast.jet.sql.impl.opt.Conventions.LOGICAL;
+
+/**
+ * Logical rule that transposes {@link DropLateItemsLogicalRel} with {@link Calc}.
+ * <p>
+ * Before:
+ * <pre>
+ * Calc[...]
+ *   DropLateItemsRel[...]
+ * </pre>
+ * After:
+ * <pre>
+ * DropLateItemsRel
+ *   Calc[...]
+ * </pre>
+ */
+@Value.Enclosing
+public class CalcDropLateItemsTransposeRule extends RelRule<RelRule.Config> implements TransformationRule {
+
+    @Value.Immutable
+    public interface Config extends RelRule.Config {
+        Config DEFAULT = ImmutableCalcDropLateItemsTransposeRule.Config.builder()
+                .description(CalcDropLateItemsTransposeRule.class.getSimpleName())
+                .operandSupplier(b0 -> b0
+                        .operand(Calc.class)
+                        .trait(LOGICAL)
+                        .inputs(b1 -> b1
+                                .operand(DropLateItemsLogicalRel.class)
+                                .anyInputs()))
+                .build();
+
+        @Override
+        default RelOptRule toRule() {
+            return new CalcDropLateItemsTransposeRule(this);
+        }
+    }
+
+    public static final RelOptRule INSTANCE = new CalcDropLateItemsTransposeRule(Config.DEFAULT);
+
+    protected CalcDropLateItemsTransposeRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        Calc calc = call.rel(0);
+        DropLateItemsLogicalRel dropRel = call.rel(1);
+
+        List<RexNode> rexNodes = calc.getProgram().expandList(calc.getProgram().getProjectList());
+        return rexNodes.stream()
+                .filter(r -> r instanceof RexInputRef)
+                .map(r -> (RexInputRef) r)
+                .anyMatch(r -> r.getIndex() == dropRel.wmField());
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        Calc calc = call.rel(0);
+        DropLateItemsLogicalRel dropRel = call.rel(1);
+        RelNode input = dropRel.getInput();
+
+        Calc newCalc = calc.copy(calc.getTraitSet(), input, calc.getProgram());
+        DropLateItemsLogicalRel newDropRel = dropRel.copy(
+                dropRel.getTraitSet(),
+                newCalc,
+                newCalc.getProgram().getSourceField(dropRel.wmField()));
```

Comment:
This is upside down. It currently does this, for example:
```
before:
-calc [projection=[$2, $0, $1]]
--drop late items [wmField=1]
---input

after:
-drop late items [wmField=0]
--calc [projection=[$2, $0, $1]]
---input
```

For `wmField` we should not use what was the source of that field before the projection in `calc`, but what is the target position of the WM field after the projection by `calc`. There's no `getTargetField()` method in the program, so we have to resort to something like we have in `matches` which I think is correct.